src/server.cpp

```diff
@@ -1018,6 +1019,39 @@ void Server::AsyncRunStep(bool initial_step)
 			m_env->saveMeta();
 		}
 	}
+
+	// Timed shutdown
+	static const float shutdown_msg_times[] =
+	{
+		1, 2, 3, 4, 5, 10, 15, 20, 25, 30, 45, 60, 120, 180, 300, 600, 1200, 1800, 3600
+	};
+
+	if (m_shutdown_timer > 0.0f) {
```

Comment:
Its the same function that calls `m_env->step`, the function which calls all the ABMs and all other lua related stuff. So no, if lua hangs, this code isn't reached either.